crates/gen/src/method.rs

```diff
@@ -135,116 +137,101 @@ impl Method {
-                            (::winrt::Interface::vtable(self).#vtable_offset)(::winrt::Abi::abi(self), #(#args)* #return_arg)
-                                .from_abi::<#return_type>(result__ )
-                        }
-                    }
-                }
-            }
-        } else {
+        // The ABI obviously still has the two composable parameters. Here we just pass the default in and out
+        // arguments to ensure the call succeeds in the non-aggregating case.
+        let composable_args = if kind == InterfaceKind::Composable {
             quote! {
-                pub fn #method_name<#constraints>(&self, #params) -> ::winrt::Result<()> {
-                    unsafe {
-                        (::winrt::Interface::vtable(self).#vtable_offset)(::winrt::Abi::abi(self), #(#args)*).ok()
-                    }
-                }
+                ::std::ptr::null_mut(), ::winrt::Abi::set_abi(&mut ::std::option::Option::<::winrt::Object>::None),
             }
-        }
-    }
-
-    pub fn gen_non_default(&self, interface: &RequiredInterface) -> TokenStream {
-        let method_name = format_ident(&self.name);
-        let params = gen_param(&self.params);
-        let constraints = gen_constraint(&self.params);
-        let args = gen_arg(&self.params);
-        let interface = interface.name.gen();
-
-        let return_type = if let Some(return_type) = &self.return_type {
-            return_type.gen_return()
         } else {
-            quote! { () }
+            TokenStream::new()
         };
 
-        quote! {
-            pub fn #method_name<#constraints>(&self, #params) -> ::winrt::Result<#return_type> {
-                <#interface as ::std::convert::From<&Self>>::from(self).#method_name(#args)
-            }
-        }
-    }
-
-    pub fn gen_static(&self, interface: &RequiredInterface) -> TokenStream {
-        let method_name = format_ident(&self.name);
-        let params = gen_param(&self.params);
-        let constraints = gen_constraint(&self.params);
-        let args = gen_arg(&self.params);
-        let interface = format_ident(&interface.name.name);
-
-        let return_type = if let Some(return_type) = &self.return_type {
+        let return_type_tokens = if let Some(return_type) = &self.return_type {
             return_type.gen_return()
         } else {
             quote! { () }
         };
 
-        quote! {
-            pub fn #method_name<#constraints>(#params) -> ::winrt::Result<#return_type> {
-                Self::#interface(|f| f.#method_name(#args))
-            }
-        }
-    }
+        let vtable_offset = Literal::u32_unsuffixed(self.vtable_offset);
 
-    pub fn gen_composable(&self, interface: &RequiredInterface) -> TokenStream {
-        let method_name = format_ident(&self.name);
-        let interface = format_ident(&interface.name.name);
+        let vcall = if let Some(return_type) = &self.return_type {
```

Comment:
Same thing as before, this might be clearer as a match:

```rust
match &self.return_type {
    Some(return_type) if return_type.array => {}
    Some(return_type) => {}
    None => {}
}
```

Reply:
I do like that but it doesn't let me share code gen for return types.